### PR TITLE
addinvoice: provide hop hints for no-amount invoice

### LINKED ISSUE #6738: addinvoice --private doesn't add hop hints if no amount is provided
### Background

`lncli addinvoice --private` returns an invoice without hop hints; if an amount is provided the `--private` option works as expected.

### Your environment

* lnd version 0.15.0-beta commit=v0.15.0-beta

### Expected behaviour

Either provide hop hints. Or if that's not technologically possible, return an error.

### PR DESCRIPTION
**Pull request by alpeb-btc**
## Change Description

This change allows creating no-amount private invoices that do provide hop hints, up to `maxHopHints`.

Fixes #6738

## Steps to Test

The `testInvoiceRoutingHints` itest was expanded to account for this use case.
If someone can provide some pointers on how to easily decode the payment request response from `lncli addinvoice`, I'd be glad to provide an example from the end-user's perspective.

This itest and the added unit test case fail before the change to `addinvoice.go` (given that change was so minimal, I clubbed it together in the same commit with the unit test change, but I can separate them if deemed necessary :slightly_smiling_face: ).

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

**Review by positiveblue — APPROVED** [review]
tACK :100:

File: lnrpc/invoicesrpc/addinvoice.go
Comment (alpeb-btc) [reply lnrpc/invoicesrpc/addinvoice.go:672]: Makes sense :+1:

File: lntest/itest/lnd_routing_test.go
Comment (alpeb-btc) [reply lntest/itest/lnd_routing_test.go:1159]: sgtm :+1:

File: lnrpc/invoicesrpc/addinvoice_test.go
Comment (joostjager) [inline lnrpc/invoicesrpc/addinvoice_test.go:679]: The test refactoring, can that go into a separate commit? That way the actual fix and functional change is observable in isolation.

File: lnrpc/invoicesrpc/addinvoice_test.go
Comment (joostjager) [inline lnrpc/invoicesrpc/addinvoice_test.go:713]: In other tests I think the written number ('two') is preferred.

File: lntest/itest/lnd_routing_test.go
Comment (joostjager) [inline lntest/itest/lnd_routing_test.go]: Refactor in separate commit helps with review.

**Comment by alpeb-btc** [comment]
Thanks @joostjager, I've rebased and rearranged commits, putting first the refactorings and typo fixes. The last three commits contain the functional change, new unit test case and new itest case, respectively.

File: lntest/itest/lnd_routing_test.go
Comment (joostjager) [inline lntest/itest/lnd_routing_test.go:1221]: Newlines removed?

File: lntest/itest/lnd_routing_test.go
Comment (joostjager) [inline lntest/itest/lnd_routing_test.go:1185]: This new structure is so much better for review, thanks!

As a final comment, I'd squash the final three commits. So in one commit, add the fix and update the tests. During review, it is nice to cross check both without going back and forth commits.

File: lntest/itest/lnd_routing_test.go
Comment (alpeb-btc) [reply lntest/itest/lnd_routing_test.go:1221]: Whoops! I'll add those back :grimacing:

File: lntest/itest/lnd_routing_test.go
Comment (alpeb-btc) [reply lntest/itest/lnd_routing_test.go:1185]: Sounds good, I've just pushed that squashing :+1:

**Comment by joostjager** [comment]
Kicked off CI - perhaps there are linter issues

**Comment by alpeb-btc** [comment]
@joostjager It appears the linter job succeeded. OTOH most of the (old?) itests are also failing in `master`. Is there a particular issue I should pay attention to?

**Comment by joostjager** [comment]
I am not sure what the status of the itests is. I see them failing a lot generally.

**Review by joostjager — APPROVED** [review]

File: lnrpc/invoicesrpc/addinvoice.go
Comment (guggero) [inline lnrpc/invoicesrpc/addinvoice.go:672]: Since this has privacy implications (zero amount potentially leaking more unannounced channels than setting a specific amount), should we add a comment to both the CLI `--amt` flag and the `value`/`value_msat` fields on the RPC?

**Review by guggero — COMMENTED** [review]
Very nice commit structure!
Just one question about documenting the behavior, other than that looks good to me.

File: lnrpc/invoicesrpc/addinvoice.go
Comment (alpeb-btc) [reply lnrpc/invoicesrpc/addinvoice.go:672]: Yep, that wouldn't hurt. I've just pushed an extra commit adding those comments into the `private` field instead, to avoid the repetition in the `value/value_msat` ones. Let me know if that works.

File: lntest/itest/lnd_routing_test.go
Comment (yyforyongyu) [inline lntest/itest/lnd_routing_test.go:1145]: Could you drop this commit? There's another ongoing big refactor #6825 which will change them altogether, thanks!

File: lntest/itest/lnd_routing_test.go
Comment (yyforyongyu) [inline lntest/itest/lnd_routing_test.go:1204]: nit: maybe this can be made into a closure inside the test since it's not used anywhere else?

File: lnrpc/invoicesrpc/addinvoice_test.go
Comment (yyforyongyu) [inline lnrpc/invoicesrpc/addinvoice_test.go:693]: strictly speaking this is testing the `consider all the open channels` part, not the `up to maxHopHints` part. I think you'll add a test case using `maxHopHints : 1` to check that.

File: lnrpc/invoicesrpc/addinvoice.go
Comment (yyforyongyu) [inline lnrpc/invoicesrpc/addinvoice.go:672]: Can we also update our itest to cover this case?

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Nice PR! Have some comments about the tests, otherwise LGTM.

File: lnrpc/invoicesrpc/addinvoice_test.go
Comment (alpeb-btc) [reply lnrpc/invoicesrpc/addinvoice_test.go:693]: Good point, I'll add that test.

**Comment by alpeb-btc** [comment]
@yyforyongyu I've pushed again addressing your feedback. LMKWYT.

File: lntest/itest/lnd_routing_test.go
Comment (yyforyongyu) [inline lntest/itest/lnd_routing_test.go:1181]: nit: we usually wrap errors, logs, in a compact mode, sth like,
```go
fmt.Errorf("unable to add "+
"invoice: %w", err)
```
This is just FYI for your future reference tho, as the itest will soon be changed so don't bother changing it here now.

File: lnrpc/invoicesrpc/addinvoice_test.go
Comment (yyforyongyu) [inline lnrpc/invoicesrpc/addinvoice_test.go:729]: 👍

File: lnrpc/invoicesrpc/addinvoice.go
Comment (yyforyongyu) [reply lnrpc/invoicesrpc/addinvoice.go:672]: actually was referring to checking the `targetAmount != 0` part, and see you already have it added here, nice!

**Review by yyforyongyu — COMMENTED** [review]
Missing a release note, and should be good to go. Before merging, maybe wait till #7126 is merged and then do a rebase to see the btcd itest build passing again.

File: lntest/itest/lnd_routing_test.go
Comment (alpeb-btc) [reply lntest/itest/lnd_routing_test.go:1181]: Good to know, thanks!

**Comment by alpeb-btc** [comment]
Thanks @yyforyongyu! I've updated the release notes in the latest commit, and I'll keep an eye open on #7126 :+1:

**Comment by guggero** [comment]
Needs a rebase because of a conflict in the release notes.

**Review by yyforyongyu — APPROVED** [review]
I think it's ready to be merged after one more rebase to fix the release note conflict, good work!

**Comment by alpeb-btc** [comment]
I've rebased and fixed the conflicts again, and the itests are a little better now :+1: :pray:

**Review by guggero — APPROVED** [review]

**Pull request merged**
